### PR TITLE
Fix pq size checking to not stop the pipeline

### PR DESCRIPTION
**Pull request by kaisecheng**
## Release notes

Change the validation of disk usage of PQ to give a warning message instead of stopping the pipeline

## What does this PR do?

This commit changes the behavior of PQ disk usage checking.
It removed the part of throwing exception that stop the pipeline.
Instead, it gives warning msg in every converge state if it fails the check

## Why is it important/What is the impact to the user?

Stopping Logstash when failing the check is considered a breaking change. User found it is too aggressive.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [ ]

## Related issues

- Fixed: #14257

### CODE REVIEW

**Comment by kaisecheng** [comment]
Q: Before we have pq size checking with multiple pipelines, there was a [BootstrapCheck](https://github.com/elastic/logstash/blob/e9c9865f4066b54048f8d708612a72d25e2fe5d9/logstash-core/lib/logstash/bootstrap_check/persisted_queue_config.rb#L23) that checked a **single** pipeline and would stop logstash if `queue.page_capacity` > `queue.max_bytes`. Reloading the pipeline did not trigger bootstrap check, hence would not stop Logstash in case bad config.  Do we want to bring back similar behavior such that any `queue.page_capacity` > `queue.max_bytes` would stop the process? 

We have done some improvements in the size checking. They are bug fixes and are considered to be not backward compatible. I certainly don't want to roll back to the old faulty one, so I think we can throw exception in this condition and give warning message for other new checking for backward compatibility

**Comment by jsvd** [comment]
There are two different kinds of checks: one is about configuration, and another is about usage.
We should stop a pipeline from starting if it has invalid settings, like a negative number of workers or, in case, a page size that is bigger than the maximum queue size.
This change should only focus on the usage check, where we want to tolerate a pipeline starting knowing that the filesystem doesn't have enough space to hold a full queue.

File: logstash-core/lib/logstash/persisted_queue_config_validator.rb [hunk 1]
```diff
@@ -74,21 +78,21 @@ def check_page_capacity(err_msg, pipeline_id, max_bytes, page_capacity)
...
     # Check disk has sufficient space for all queues reach their max bytes. Queues may config with different paths/ devices.
     # It uses the filesystem of the path and count the required bytes by filesystem
-    def check_disk_space(err_msg, queue_path_file_system, required_free_bytes)
-      disk_err_msg =
+    def check_disk_space(warn_msg, queue_path_file_system, required_free_bytes)
+      disk_warn_msg =
         queue_path_file_system
           .select { |queue_path, file_system| !FsUtil.hasFreeSpace(Paths.get(queue_path), required_free_bytes.fetch(file_system, 0)) }
-          .map { |queue_path, file_system| "Persistent queue path #{queue_path} is unable to allocate #{required_free_bytes.fetch(file_system, 0)} more bytes on top of its current usage." }
+          .map { |queue_path, file_system| "Persistent queue #{queue_path} in #{file_system} is unable to allocate #{required_free_bytes.fetch(file_system, 0)} more bytes on top of its current usage." }
```
Comment (jsvd) [inline logstash-core/lib/logstash/persisted_queue_config_validator.rb:93]:
```suggestion
          .map { |queue_path, file_system| "The persistent queue on path \"#{queue_path}\" won't fit in file system \"#{file_system}\" when full. Please free or allocate #{required_free_bytes.fetch(file_system, 0)} more bytes." }
```

**Review by jsvd — APPROVED** [review]
Tested locally and works as expected. I suggested a tweak to the warning message, otherwise LGTM

**Comment by kaisecheng** [comment]
@logstashmachine backport 7.17

**Comment by kaisecheng** [comment]
@logstashmachine backport 8.3

**Pull request merged**